pkg/websocket/client.go

```diff
@@ -437,7 +440,9 @@ func (c *Client) writePump() {
 				if ws.IsUnexpectedCloseError(err, ws.CloseNormalClosure) {
 					c.cfg.Log.Error("write error: ", err)
 				}
-				c.notifyClose <- err
```

Comment:
The most reliable way to handle something like this is to use a select over sending and a shutdown signal (or whatever signals that nothing will be reading the other side). In this case, there's [`c.stopWritePump`](https://github.com/stripe/stripe-cli/blob/aed27b56a08c690f9ca1a83dcbecda5e59ef5628/pkg/websocket/client.go#L449), which seems to be closed and re-initialized when the connection is reset.

So perhaps something like this would resolve the deadlock?

```go
select {
case c.notifyClose <- err:
case <-c.stopWritePump:
}
```

I think there might still be a race reading `stopWritePump` from the client struct, however, since it appears to be swapped out without locking or coordination. I might suggest, then, also updating `go c.writePump()` to `go c.writePump(c.stopWritePump)` so writePump can always read the correct instance of the channel from its arguments.

Reply:
Thanks for this Bernerd! This is the method we went with.